pkg/loadbalancerservice/controller.go

```diff
@@ -0,0 +1,205 @@
+func (c *LoadbalancerServiceController) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
+	defer close(stopped)
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	config, err := c.restConfig()
+	if err != nil {
+		return errors.Wrap(err, "error creating rest config for service controller")
+	}
+	c.client, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "failed to create clientset for service controller")
+	}
+
+	klog.Infof("Starting service controller")
+
+	factory := informers.NewSharedInformerFactory(c.client, defaultInformerResyncPeriod)
+	serviceInformer := factory.Core().V1().Services()
+	c.informer = serviceInformer.Informer()
+	c.queue = workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	c.indexer = c.informer.GetIndexer()
+	c.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
+	})
+
+	factory.Start(stopCh)
+
+	if !cache.WaitForCacheSync(stopCh, c.informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
+	}
+
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	close(ready)
+
+	<-ctx.Done()
+
+	return ctx.Err()
+}
+
+func (c *LoadbalancerServiceController) runWorker() {
+	for c.processNextItem() {
+	}
+}
+
+func (c *LoadbalancerServiceController) processNextItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.updateServiceStatus(key.(string))
+	c.handleErr(err, key)
+	return true
+}
+
+// handleErr checks if an error happened and makes sure we will retry later.
+func (c *LoadbalancerServiceController) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	klog.Infof("Error syncing service %v: %v", key, err)
+
+	// Re-enqueue the key rate limited. Based on the rate limiter on the
+	// queue and the re-enqueue history, the key will be processed later again.
+	c.queue.AddRateLimited(key)
+}
+
+func (c *LoadbalancerServiceController) updateServiceStatus(key string) error {
+	obj, exists, err := c.indexer.GetByKey(key)
+	if err != nil {
+		klog.Errorf("Fetching service object with key %s from store failed with %v", key, err)
+		return err
+	}
+
+	if !exists {
+		klog.Infof("Service %s does not exist anymore", key)
+	} else {
+		svc := obj.(*corev1.Service)
+		if svc.Spec.Type != corev1.ServiceTypeLoadBalancer {
+			return nil
+		}
+		klog.Infof("Process service %s/%s", svc.Namespace, svc.Name)
+
+		newStatus := c.getNewStatus(svc)
+		if len(newStatus.Ingress) == 0 {
```

Comment:
I feel like this is too implicit. Can `getNewStatus()` return `(*corev1.LoadBalancerStatus, error)` and error would happen in this place:
```go
				if ep.Port == np.Port {
					klog.Infof("Node port %d occupied", ep.Port)
					return newStatus
				}
```
new:
```go
				if ep.Port == np.Port {
					return nil, fmt.Error("Node port %d occupied", ep.Port)
				}
```

Reply:
done